### PR TITLE
HPCC-21392 Prevent automatic translation if file types mismatch

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
NB: this change is for hthor and thor, I have opened a separate JIRA for roxie changes: HPCC-21745

**Comment by jakesmith** [comment]
@ghalliday - please review

**Comment by jakesmith** [comment]
@ghalliday - this should probably target 7.0.x ?

**Comment by richardkchapman** [comment]
No change in Roxie?

**Comment by jakesmith** [comment]
>NB: this change is for hthor and thor, I have opened a separate JIRA for roxie changes: HPCC-21745

separate issue was opened.
Suspect similar change needed, but kept separate partly because wasn't sure where.

**Comment by jakesmith** [comment]
Retargeted at 7.0.x

@mckellyln - please review.

**Comment by jakesmith** [comment]
@mckellyln - rebasing to 7.0.x caused build issue (due to missing stricmp). Rectified now.

File: ecl/hthor/hthor.cpp [hunk 1]
```diff
@@ -8064,16 +8064,21 @@ void CHThorDiskReadBaseActivity::ready()
     if (ldFile)
         dFile = ldFile->queryDistributedFile();  // Null for local file usage
 
-    Linked<IOutputMetaData> publishedMeta = actualDiskMeta;
-    unsigned publishedCrc = actualCrc;
+    Owned<IOutputMetaData> publishedMeta;
```
Comment (mckellyln) [inline ecl/hthor/hthor.cpp]: What is publishedMeta initialized to ?  Previously it was to actualDiskMeta.

Comment (jakesmith) [reply ecl/hthor/hthor.cpp]: It is not.
It is intentionally left empty, unless format matches

File: thorlcr/activities/diskread/thdiskreadslave.cpp [hunk 2]
```diff
@@ -68,14 +68,24 @@ class CDiskReadSlaveActivityRecord : public CDiskReadSlaveActivityBase, implemen
...
+        Linked<IOutputMetaData> publishedFormat;
+        unsigned publishedFormatCrc = 0;
+
+        const char *kind = queryFileKind(&partDesc.queryOwner());
+        if (strisame(kind, "flat") || (RecordTranslationMode::AlwaysDisk == translationMode))
+        {
+            IPropertyTree const &props = partDesc.queryOwner().queryProperties();
+            publishedFormat.setown(getDaliLayoutInfo(props));
+            publishedFormatCrc = (unsigned)props.getPropInt("@formatCrc", 0);
+        }
```
Comment (mckellyln) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: If this isn't true should published be set equal to expected ?

**Comment by mckellyln** [comment]
@jakesmith looks good, but have two questions above.

File: thorlcr/activities/diskread/thdiskreadslave.cpp (hunk 2 above)
Comment (jakesmith) [reply thorlcr/activities/diskread/thdiskreadslave.cpp]: no, it's correct to be 0, if missing.
It won't make any difference, but should not set even if @formatCrc is present if 'publishedFormat' is not set. Will change.

**Comment by jakesmith** [comment]
@mckellyln - see change above.

**Review by mckellyln — APPROVED** [review]

**Comment by mckellyln** [comment]
@jakesmith looks good.

**Comment by jakesmith** [comment]
@richardkchapman - squashed, please merge.

**Pull request merged**